cpp/oneapi/dal/backend/interop/homogen_table_adapter.hpp

```diff
@@ -0,0 +1,350 @@
+            return status_t(daal::services::ErrorIncorrectIndex);
+        }
+
+        try {
+            array<BlockData> values;
+            auto block_ptr = block.getBlockPtr();
+            if (block_ptr != nullptr && info.size >= info.row_count) {
+                values.reset(block_ptr, info.size, dal::empty_delete<BlockData>());
+            }
+
+            column_accessor<const BlockData> acc{ original_table_ };
+            if (block_ptr != pull_values(acc,
+                                         values,
+                                         info.column_index,
+                                         range{ info.row_begin_index, info.row_end_index })) {
+                setBlockData(values, info.row_count, 1, block);
+            }
+        }
+        catch (const std::exception&) {
+            return status_t(daal::services::UnknownError);
+        }
+
+        return status_t();
+    }
+
+    bool check_row_indexes_in_range(const block_info& info) {
+        const std::int64_t row_count = original_table_.get_row_count();
+        return info.row_begin_index >= 0 && info.row_begin_index < row_count &&
+               info.row_end_index > info.row_begin_index && info.row_end_index <= row_count;
+    }
+
+    bool check_column_index_in_range(const block_info& info) {
+        const std::int64_t column_count = original_table_.get_column_count();
+        return info.column_index >= 0 && info.column_index < column_count;
+    }
+
+    template <typename BlockData>
+    void setBlockData(array<BlockData>& data,
+                      std::int64_t row_count,
+                      std::int64_t column_count,
+                      block_desc_t<BlockData>& block) {
+        auto raw_ptr = const_cast<BlockData*>(data.get_data());
+        if constexpr (is_host_only) {
+            auto data_shared =
+                daal::services::SharedPtr<BlockData>(raw_ptr, daal_array_owner(data));
+            block.setSharedPtr(data_shared, column_count, row_count);
+        }
+#ifdef ONEAPI_DAL_DATA_PARALLEL
+        else {
+            daal::services::Buffer<BlockData> buffer(raw_ptr,
+                                                     row_count * column_count,
+                                                     sycl::usm::alloc::shared);
+            block.setBuffer(buffer, column_count, row_count);
+        }
+#endif
+    }
+
+    template <typename Accessor, typename... Args>
+    auto pull_values(const Accessor& acc, Args&&... args) {
+        if constexpr (is_host_only) {
+            return acc.pull(std::forward<Args>(args)...);
+        }
+#ifdef ONEAPI_DAL_DATA_PARALLEL
+        else {
+            return acc.pull(policy_.get_queue(),
+                            std::forward<Args>(args)...,
+                            sycl::usm::alloc::shared);
+        }
+#endif
+    }
+
+    homogen_table_adapter(const Policy& policy, const homogen_table& table, status_t& stat)
+            : base(const_cast<Data*>(table.get_data<Data>()),
```

Comment:
I see potential problem here.

Implementation of old homogen table assumes that the shared pointer to the data can be extracted from the table and then table can be safely destroyed, but shared pointer will be valid (see method `getArraySharedPtr`).

A call of `base` with the provided arguments causes creation of shared pointer with empty deleter inside `HomogenNumericTable`, so the case mentioned above will not be supported and will result to dangling pointer.

This case is not likely to happen, but if it happen we will not know about it, there is no way to throw exception/assert.

So, I suggest to wrap data pointer with the shared pointer that holds `original_table_` in deleter or convert `std::shared_ptr` to `SharedPtr` and invoke `base` constructor that accepts shared pointer.